README.md
# Transfer-attacks-comparison
Comparing State of the art black-box attacks methods such as MI, DI, Admix, Ghost networks by attacking some robust models trained on cifar10 that are available in the robustbench leaderboad.

## Reproducing the results
Download pretrained models on cifar-10 from : https://github.com/huyvnphan/PyTorch_CIFAR10.git 
- Run the following command to get correctly classified images by all sources and targets models
```bash
python data.py --sources Resnet50 --targets 1 2 3 61 --n_images 100
```
 by doing that, you discard images that were already misclassified by the targets model. The ids of images will be stored in a config file



- Run this command to evaluate the attack (`TI-FGSM.py`, `DI-FGSM.py`, `adm.py`)  
The ids of correctly classified test images will be loaded from config file generated by `data.py`
```bash
python MI-FGSM.py
```

## Results

The following results are obtained by using the default parameters of each attack, no tuning have been made.the top 5 rows are the robust accuracies of the top five robust models and the last row is the robust accuracy of the vanilla model.

### Query efficient meta attack
### distance: l2
### num_images: 100 

Model | rank | Query efficient meta attack |
----------- | --- | ---------
Gowal2020Uncovering_70_16_extra | 3 | 0.84 |
Gowal2021Improving_28_10_ddpm_100m | 7 | 0.35 | 
Carmon2019Unlabeled | 23 | 0.3 |


### Query efficient meta attack
### distance: linf
### num_images: 100 

Model | rank | Query efficient meta attack |
----------- | --- | ---------
Gowal2020Uncovering_70_16_extra | 3 | 0.81 |
Gowal2021Improving_28_10_ddpm_100m | 7 | 0.35 | 


### Generalized Data-free Universal Adversarial Perturbations
### eps = 8/255
### num_images: 100 


UAP 2       |
----------- |
0.92        |
0.89        |   
0.93        |
0.94        |
0.94        |
0.61        |



### Source: Resnet50
### eps = 8/255
### num_images: 100 

MI-FGSM  |  TI-FGSM  |  VMI-FGSM  |  VNI-FGSM  |  GN-MIFGSM  |  DI-FGSM
---------|-----------|------------|------------|-------------|---------
1.0      |  1.0      |  1.0       |  1.0       |  0.99       |  1.0
1.0      |  1.0      |  1.0       |  1.0       |  1.0        |  1.0
1.0      |  1.0      |  1.0       |  1.0       |  1.0        |  1.0
1.0      |  1.0      |  1.0       |  1.0       |  0.99       |  1.0
1.0      |  0.98     |  1.0       |  1.0       |  0.99       |  1.0
0.43     |  0.91     |  0.3       |  0.33      |  0.72       |  0.83


### Source: Resnet50
### eps = 16/255
### num_images: 100

MI-FGSM  |  VMI-FGSM  |  TI-FGSM  |  GN-MIFGSM  |  DI-FGSM  |  VNI-FGSM
---------|------------|-----------|-------------|-----------|----------
0.98     |  0.97      |  0.98     |  0.98       |  0.99     |  0.97
0.98     |  0.97      |  0.96     |  0.98       |  0.99     |  0.98
0.96     |  0.96      |  0.97     |  0.99       |  0.99     |  0.96
0.97     |  0.96      |  0.95     |  0.97       |  0.99     |  0.97
0.97     |  0.96      |  0.94     |  0.97       |  0.99     |  0.96
0.18     |  0.06      |  0.68     |  0.4        |  0.65     |  0.05


### Source: Resnet50
### eps = 32/255
### num_images: 100 

MI-FGSM  |  DI-FGSM  |  TI-FGSM  |  GN-MIFGSM  |  VMI-FGSM  |  VNI-FGSM
---------|-----------|-----------|-------------|------------|----------
0.95     |  0.99     |  0.95     |  0.98       |  0.97      |  0.97
0.96     |  0.99     |  0.93     |  0.97       |  0.98      |  0.98
0.96     |  1.0      |  0.96     |  0.99       |  0.97      |  0.98
0.95     |  0.99     |  0.9      |  0.96       |  0.98      |  0.98
0.95     |  0.99     |  0.9      |  0.96       |  0.98      |  0.98
0.15     |  0.64     |  0.58     |  0.34       |  0.1       |  0.03


### Source: Resnet50
### num_images: 100 

Admix       |
----------- |
1.0         |
1.0         |   
1.0         |
0.99        |
0.99        |
0.25        |


### Source: Model rank 7
### eps = 8/255
### num_images: 100 

DI-FGSM  |  MI-FGSM  |  TI-FGSM  |  VMI-FGSM  |  VNI-FGSM
---------|-----------|-----------|------------|----------
0.86     |  0.88     |  0.97     |  1.0       |  1.0
0.85     |  0.85     |  0.9      |  0.99      |  1.0
0.82     |  0.82     |  0.93     |  0.99      |  0.99
0.82     |  0.82     |  0.89     |  0.97      |  0.98
0.81     |  0.81     |  0.86     |  0.99      |  0.99
0.92     |  0.88     |  0.96     |  0.93      |  0.94


### Source: Model rank 7
### eps = 16/255
### num_images: 100 

DI-FGSM  |  MI-FGSM  |  TI-FGSM  |  VMI-FGSM  |  VNI-FGSM
---------|-----------|-----------|------------|----------
0.61     |  0.63     |  0.78     |  0.95      |  0.95
0.55     |  0.58     |  0.74     |  0.89      |  0.87
0.57     |  0.61     |  0.72     |  0.9       |  0.86
0.59     |  0.59     |  0.73     |  0.85      |  0.85
0.59     |  0.61     |  0.73     |  0.86      |  0.85
0.59     |  0.52     |  0.89     |  0.75      |  0.78


### Source: Model rank 7
### eps = 32/255
### num_images: 100 

DI-FGSM  |  MI-FGSM  |  TI-FGSM  |  VMI-FGSM  |  VNI-FGSM
---------|-----------|-----------|------------|----------
0.5      |  0.52     |  0.69     |  0.89      |  0.91
0.44     |  0.46     |  0.64     |  0.91      |  0.92
0.49     |  0.49     |  0.66     |  0.87      |  0.88
0.48     |  0.48     |  0.66     |  0.89      |  0.89
0.48     |  0.49     |  0.66     |  0.89      |  0.89
0.4      |  0.36     |  0.82     |  0.68      |  0.69


### Source: Carmon19
### eps = 16/255
### num_images: 100 

DI-FGSM  |  GN-MIFGSM  |  MI-FGSM  |  TI-FGSM  |  VMI-FGSM  |  VNI-FGSM
---------|-------------|-----------|-----------|------------|----------
0.62     |  0.63       |  0.97     |  0.99     |  0.95      |  0.95
0.67     |  0.71       |  0.96     |  0.96     |  0.91      |  0.92
0.57     |  0.59       |  0.95     |  0.98     |  0.92      |  0.92
0.6      |  0.64       |  0.93     |  0.92     |  0.9       |  0.9
0.63     |  0.67       |  0.93     |  0.93     |  0.91      |  0.91
0.56     |  0.54       |  0.82     |  0.95     |  0.79      |  0.82

### Source: Vanilla
### eps = 16/255
### num_images: 100 

DI-FGSM  |  GN-MIFGSM  |  MI-FGSM  |  TI-FGSM  |  VNI-FGSM
---------|-------------|-----------|-----------|----------
1.0      |  0.99       |  0.99     |  1.0      |  0.99
1.0      |  0.99       |  1.0      |  0.99     |  1.0
0.99     |  1.0        |  1.0      |  0.99     |  1.0
0.99     |  1.0        |  1.0      |  0.99     |  1.0
0.99     |  1.0        |  1.0      |  0.97     |  1.0
0.0      |  0.0        |  0.33     |  0.96     |  0.3


### MI FGSM ensemble
#### Source: Resnet50, Densenet169, Vgg19
#### num_images: 100

MI-FGSM ENS |
----------- |
1.0         |
1.0         |   
1.0         |
1.0         |
1.0         |
0.22        |


#### Source: Resnet50, Resnet18
#### num_images: 100

MI-FGSM ENS |
----------- |
1.0         |
1.0         |
1.0         |
1.0         |
1.0         |   
0.24        |


### Dast
#### Source: vgg
#### num_images: 100

| <sub>Model ID</sub> | <sub> Dast Robust accuracy</sub> |
|:---:|:---:|
| <sub>**Carmon2019Unlabeled**</sub> | <sub>99%</sub> |
| <sub>**Standard**</sub> | <sub>56%</sub> |




### TREMBA 

#### Source: resnet50
#### num_images: 100

| <sub>Model ID</sub> | <sub> Tremba Robust accuracy</sub> |
|:---:|:---:|
| <sub>**Carmon2019Unlabeled**</sub> | <sub>100%</sub> |
| <sub>**Standard**</sub> | <sub>64%</sub> |



BASES/changes.TXT
arguments


source
targets

dataloader

loop of attack
remove var: gt_label_name, tgt_labels and var commented
remove imagenet_names[label_idx]
remove save adv
remove plots

util_bases
comment normaalize
remove permute


data.py
import argparse
import torch
import torch.nn as nn
from robustbench.utils import load_model
from robustbench.data import load_cifar10
from cifar10_models.resnet import *
from cifar10_models.densenet import *
from cifar10_models.vgg import *

import math
import json
from Normalize import Normalize

#add:
    # device for gpu
    #error: resnet50 pth

#future:
    #cross validation

# modified robustbench.utils
def correct_predict_ids(models,
                   x: torch.Tensor,
                   y: torch.Tensor,
                   n_images,
                   log_interval,
                   print_log,
                   batch_size: int = 100,
                   device: torch.device = None):
    if device is None:
        device = x.device
    acc = 0.
    ids = []
    n_batches = math.ceil(x.shape[0] / batch_size)
    with torch.no_grad():
        for counter in range(n_batches):
            x_curr = x[counter * batch_size:(counter + 1) *
                       batch_size].to(device)
            y_curr = y[counter * batch_size:(counter + 1) *
                       batch_size].to(device)
            
            accs_models= torch.empty((len(models),batch_size), dtype=torch.float).to(device) #rmv
            for i,model in enumerate(models):
                output = model(x_curr)
                accs_models[i,:] = (output.max(1)[1] == y_curr).float()
            
            ids_batch = torch.nonzero(accs_models.bool().all(dim=0))
            ids.extend( (ids_batch + counter * batch_size).
                        view(-1).tolist()
            )
            if (counter+1) % log_interval == 0 and print_log:
                print('{} elements selected from batch {}, total of {}'.format(torch.numel(ids_batch),counter+1,len(ids)))

            if len(ids) >= n_images:
                break

    if len(ids) < n_images:
        print('Only ',len(ids),' correctly classified images by all target models')
    
    return ids[:n_images]
    

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    #select the first n images correctly classified by all targeted models
    parser.add_argument('--n_images', type=int, default=100 , help = 'number of test images')
    parser.add_argument('--batch_size', type=int, default=20 , help = 'batch size')
    parser.add_argument('--sources', nargs = '+', default=['Resnet50','Resnet18'], help = 'source model')
    parser.add_argument('--targets', type=int,nargs = '+', default = [1, 3, 61], help = 'ranks of target models from cifar-10 linf, eps=2/288 robustbench leaderboad')
    parser.add_argument('--log_interval', default= 1, help = 'print each n batch')
    parser.add_argument('--print_log', default=True, help = 'print log in terminal')
    parser.add_argument('--output', default='config_ids_source_targets', help = 'file name contains the correctly classified images plus the source and target models')#output dir or filename

    args = parser.parse_args()

    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

    mean, std  = (0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)
    source_models = []
    for source in args.sources:
        if source == 'Resnet50': 
            source_model = resnet50(pretrained=True)
        elif source == 'Resnet34': 
            source_model = resnet34(pretrained=True)
        elif source == 'Resnet18':
            source_model = resnet18(pretrained=True)
        elif source == 'Densenet169': 
            source_model = densenet169(pretrained=True)
        elif source == 'Densenet161': 
            source_model = densenet161(pretrained=True)
        elif source == 'Densenet121': 
            source_model = densenet121(pretrained=True)
        elif source == 'Vgg19':
            source_model = vgg19_bn(pretrained=True)
        elif source == 'Resnet50_non_normalize': 
            source_model = resnet50(pretrained=False)
            PATH = 'resnet50_128_100'
            source_model.load_state_dict(torch.load(PATH,map_location=device))
        else :
            raise ValueError('Source model is not recognizable')
        
        source_model = nn.Sequential(
            Normalize(mean, std),
            source_model
        )
        #todevice
        source_model.eval() 
        source_models.append(source_model)


    rank_to_id = {
        1 : 'Rebuffi2021Fixing_70_16_cutmix_extra',
        2 : 'Gowal2021Improving_70_16_ddpm_100m',
        3 : 'Gowal2020Uncovering_70_16_extra',
        4 : 'Rebuffi2021Fixing_106_16_cutmix_ddpm',
        5 : 'Rebuffi2021Fixing_70_16_cutmix_ddpm',
        23 :'Carmon2019Unlabeled',
        61 : 'Standard',
    }

    target_models = [] 
    rank_to_id_selected = dict()
    for rank in args.targets:
        if rank in rank_to_id.keys():
            print('Loading model ', rank)
            target_model = load_model(model_name=rank_to_id[rank], dataset='cifar10', threat_model='Linf')
            #todevice
            target_models.append(target_model)
            rank_to_id_selected[rank] = rank_to_id[rank]
        else: 
            raise ValueError('target model with ranking = ',rank,' is not available')

    x_test, y_test = load_cifar10()
    
    ids = correct_predict_ids(target_models+source_models, x_test, y_test, args.n_images, args.log_interval, args.print_log, args.batch_size) # + source_models get correct  predict images from source models

    config = {
        'ids' : ids,
        'sources' : args.sources,
        'targets' : rank_to_id_selected,
    }


    with open(args.output + '.json','w') as f:
        json.dump(config, f, indent=4)